scylla/src/transport/load_balancing/default.rs

```diff
@@ -283,13 +309,19 @@ impl DefaultPolicy {
         predicate: impl Fn(&NodeRef<'a>) -> bool,
         cluster: &'a ClusterData,
     ) -> impl Iterator<Item = NodeRef<'a>> {
+        let predicate = move |node| match replica_location {
+            ReplicaLocation::Any | ReplicaLocation::Datacenter => predicate(&node),
+            ReplicaLocation::DatacenterAndRack => {
+                predicate(&node) && node.rack == self.preferred_rack
+            }
+        };
```

Comment:
I think that it belongs to the previous commit, where you added `ReplicaLocation` parameter to `pick_replica`, `replicas` and `shuffled_replicas`. 

This is because without this predicate modification, those functions are not fully respecting `ReplicaLocation` values - `replicas(..., ReplicaLocation::DatacenterAndRack, ...)` won't return a set that is restricted to local rack.